Implement Paginator sortableFields builder and combined sorting key.

dereuromark opened the pull request:
Resolves https://github.com/cakephp/cakephp/issues/18897

Feel free to further adjust.

dereuromark committed:
Paginator sortMap.

josbeir commented:
Looks good, should we handle  what i mentioned here: https://github.com/cakephp/cakephp/issues/18897#issuecomment-3270568455

(sharing the same field name on multiple sortmaps)

dereuromark commented:
I don't think we need validation to prevent "overlapping" fields since each mapped key represents a complete, independent sorting strategy. This is actually a feature - it lets you create multiple sort options that include
   the same fields in different combinations.

  The concern about ambiguity is resolved by the design:
  - No ambiguity exists because direct field sorting is disabled when sortMap is defined
  - Each sort key (item1, item2) is self-contained and independent
  - The question "In what direction should title be sorted?" doesn't arise because title itself is not a valid sort key

josbeir commented:
> I don't think we need validation to prevent "overlapping" fields since each mapped key represents a complete, independent sorting strategy. This is actually a feature - it lets you create multiple sort options that include the same fields in different combinations.
> 
> The concern about ambiguity is resolved by the design:
> 
>     * No ambiguity exists because direct field sorting is disabled when sortMap is defined
> 
>     * Each sort key (item1, item2) is self-contained and independent
> 
>     * The question "In what direction should title be sorted?" doesn't arise because title itself is not a valid sort key

Hmm, i'm probably not understanding it completely but  

```php
'sortMap' => [
  'item1' => ['title', 'created']
  'item2' => ['title', 'modified']
]

// sorting on both definitions: 
&item1=asc&item2=desc
```

Translates into the following statement right ?

```sql
SORT BY
   title ASC
   created ASC,
   title DESC,
   modified DESC 
```

ADmad commented:
> &item1=asc&item2=desc

The paginatorhelper doesn't generate query strings like this. The direction can be specific only for 1 field, `sort=item1&dir=asc`

dereuromark commented:
Imo we dont even need even sort asc/desc if we bind that into the key.

    sort=title-asc
    sort=title-desc

A flat list with all possible options

ADmad commented:
> Imo we dont even need even sort asc/desc if we bind that into the key.

Yes that could be an additional enhancement, though what I would really like is the ability to do multi field/colum sort. The query string for it could be like `sort[f1]=asc&sort[f2]=desc`

josbeir commented:
@ADmad I think that kind of sort style is not very readable for the end user. This is something Drupal (facets/views) does and it is constantly a fight to get right for proper SEO

Maybe something like `/products?sort=price:asc,rating:desc` would be more friendly ?
or `/products?sort=+price,-rating`

With that said, we should of course be weary not to go back to Cake 1-2.x style query params. Good times... :melting_face: 

Another thought that crosses my mind is then making sure that it is always returned in the same order to keep a canonical structure.

ADmad commented:
> I think that kind of sort style is not very readable for the end user

The average non-techie user doesn't read/understand query string params :)

> This is something Drupal (facets/views) does and it is constantly a fight to get right for proper SEO

I am not knowledgeable enough about how search engines deal with query string these days to be able to comment on this.

> Maybe something like /products?sort=price:asc,rating:desc would be more friendly

This could work I guess.

BTW if for e.g. you have a search form with a multi-value field and using PRG you will get `foo[]` in the query string :)

markstory commented:
> Translates into the following statement right ?

Yes. You ask a bad question, you get a bad answer. The SQL you posted is syntactically correct, and is what I would expect to happen if all of those ordering clauses were combined. 

> A flat list with all possible options

This could work but we'll need to make `PaginatorHelper` aware of how to change direction. How would application code express that `field-asc` and `field-desc` are related and opposite directions?

> With that said, we should of course be weary not to go back to Cake 1-2.x style query params. Good times... 🫠

No thank you :smile:

markstory commented on src/Datasource/Paging/NumericPaginator.php:
```diff
@@ -49,6 +49,22 @@ class NumericPaginator implements PaginatorInterface
      *   sorting on either associated columns or calculated fields then you will
      *   have to explicitly specify them (along with other fields). Using an empty
      *   array will disable sorting alltogether.
+     * - `sortMap` - A map of sort keys to their corresponding database fields. Allows
+     *   creating friendly sort keys that map to one or more actual fields. When defined,
+     *   only the mapped keys will be sortable. Supports simple mapping, multi-column
+     *   sorting, and fixed direction sorting. You can also use numeric arrays for 1:1
+     *   mappings where the field name is the same as the sort key. Example:
```
This has some overlap with `sortableFields` should we keep both long term?

ADmad replied:
`sortableFields` does become redundant with the addition of `sortMap`, so it could be deprecated.

markstory replied:
Yes, I think we can make this into something that is just as simple to use and handles more advanced use cases as well.

dereuromark replied:
Isnt it already now? :)

dereuromark committed:
Showcasing combined sort-dir.

dereuromark commented:
I added a demo of combined sort & dir.
Wonder if that would work out better.

dereuromark commented:
For me it looks quite complete.

 What's Missing:

  A way to specify default directions that can still be toggled. For example:
  - Default: title ASC, created DESC
  - When toggled: title DESC, created ASC

  Potential Solution:

  We could enhance the syntax to support default directions that flip together:
```php
  'sortMap' => [
      // New syntax: use '@' prefix for toggleable fields with defaults
      'newest' => [
          '@title' => 'asc',     // Default asc, flips to desc
          '@created' => 'desc',  // Default desc, flips to asc
      ],
  ]
```
  When user clicks the sort link:
  - First click: title ASC, created DESC (defaults)
  - Second click: title DESC, created ASC (inverted)

We could also invert this and make the @ signal that this is hardcoding the direction, as this might be more rare used.
Or, thinking about it (and in line of other languages, e.g. CSS), 

      'newest' => [
          'title' => 'asc!',     // Only asc
          'created' => 'desc',  // Default desc, flips to asc
      ],

Use ! to signal that this is hardcoded in that direction.

josbeir commented:
> For me it looks quite complete.
> 
> What's Missing:
> 
> A way to specify default directions that can still be toggled. For example:
> 
>     * Default: title ASC, created DESC
> 
>     * When toggled: title DESC, created ASC
> 
> 
> Potential Solution:
> 
> We could enhance the syntax to support default directions that flip together:
> 
> ```
>   'sortMap' => [
>       // New syntax: use '@' prefix for toggleable fields with defaults
>       'newest' => [
>           '@title' => 'asc',     // Default asc, flips to desc
>           '@created' => 'desc',  // Default desc, flips to asc
>       ],
>   ]
> ```
> 
> When user clicks the sort link:
> 
>     * First click: title ASC, created DESC (defaults)
> 
>     * Second click: title DESC, created ASC (inverted)
> 
> 
> We could also invert this and make the @ signal that this is hardcoding the direction, as this might be more rare used. Or, thinking about it (and in line of other languages, e.g. CSS),
> 
> ```
>   'newest' => [
>       'title' => 'asc!',     // Only asc
>       'created' => 'desc',  // Default desc, flips to asc
>   ],
> ```
> 
> Use ! to signal that this is hardcoded in that direction.

I like the idea!

rochamarcelo commented:
Could we have classes to define how a field should be ordered instead of a new syntax?  

```php

class OrderField 
{
  
    public function defaultDir(): ?string
    {
    }

    public function isFixedDir(): bool
    {
    }

    public function ormColumn(): array
}
```

dereuromark commented:
As opt-in? I feel like as default this could become quite some overhead for apps to create.

dereuromark committed:
Use ! for locking.

dereuromark commented:
or do you mean sth like
```php
 namespace Cake\Datasource\Paging;

  class SortField 
  {
      protected string $field;
      protected ?string $defaultDirection;
      protected bool $locked;

      public function __construct(string $field, ?string $defaultDirection = null, bool $locked = false)
      {
          $this->field = $field;
          $this->defaultDirection = $defaultDirection;
          $this->locked = $locked;
      }

      public static function asc(string $field): self
      {
          return new self($field, 'asc', false);
      }

      public static function desc(string $field): self
      {
          return new self($field, 'desc', false);
      }

      public static function locked(string $field, string $direction): self
      {
          return new self($field, $direction, true);
      }

      public function getField(): string
      {
          return $this->field;
      }

      public function getDirection(string $requestedDirection, bool $directionSpecified): string
      {
          if ($this->locked) {
              return $this->defaultDirection;
          }

          if (!$directionSpecified && $this->defaultDirection) {
              return $this->defaultDirection;
          }

          return $requestedDirection;
      }

      public function isLocked(): bool
      {
          return $this->locked;
      }
  }
```

  Usage Examples:
```php
  'sortMap' => [
      'newest' => [
          SortField::desc('created'),  // Default desc, toggleable
          SortField::asc('title'),     // Default asc, toggleable
      ],
      'popular' => [
          SortField::locked('score', 'desc'),  // Always desc
          'author',  // Still support strings for BC
      ],
  ]
```

rochamarcelo commented:
Yes, I like that. Maybe add an interface for it with  methods `getField`  and `isLocked`.

I'm not sure if the methods `desc` and `asc` are necessary, but the names could be clear, something like defaultDesc  and defaultAsc.

josbeir commented:
We could maybe also create factory class for building the sortmap then. This would allow for a more strict and readable way for defining them.

```php
$factory = SortMapFactory::create()
    ->field('title')
    ->field('created')
    ->field('rank', locked: true, default: 'desc')
    ->combo('newest', [
        SortField::desc('created'),
        SortField::asc('title'),
    ])
    ->combo('popular', [
        SortField::desc('score'),
        SortField::asc('comments_count'),
    ])
    ->combo('recently-updated', [
        SortField::desc('modified'),
        SortField::asc('title'),
    ])
    ->raw('alpha-group', [
        'group_name' => 'asc',
        'name' => 'asc',
    ]);

dereuromark committed:
Add factory and object approach.

dereuromark commented:
I added factory and SortField now with tests.
Also deprecated old sortableFields now.

dereuromark committed:
Add factory and object approach.

dereuromark committed:
Add factory and object approach.

dereuromark committed:
Add factory and object approach.

dereuromark committed:
Merge branch '5.next' into 5.next-paginator-sort-map

markstory commented on src/Datasource/Paging/NumericPaginator.php:
```diff
@@ -668,6 +736,92 @@ protected function _prefix(RepositoryInterface $object, array $order, bool $allo
+        if (is_array($mapping)) {
+            foreach ($mapping as $key => $value) {
+                // Handle SortField objects
+                if ($value instanceof SortField) {
+                    $field = $value->getField();
+                    $fieldDirection = $value->getDirection($direction, $directionSpecified);
+                    $order[$field] = $fieldDirection;
+                } elseif (is_int($key)) {
+                    // Indexed array: field uses querystring direction
+                    // e.g., ['modified', 'name']
+                    $order[$value] = $direction;
+                } elseif (str_ends_with($value, '!')) {
+                    // Associative array: check for locked (!) or default direction
+                    // Locked direction (ends with !): always use specified direction
+                    // e.g., ['created' => 'desc!'] always sorts desc
+                    $order[$key] = rtrim($value, '!');
+                } elseif (!$directionSpecified) {
+                    // Default direction that can be toggled
+                    // No direction specified, use the default
+                    $order[$key] = $value;
+                } else {
+                    // Direction specified, use it for all toggleable fields
+                    $order[$key] = $direction;
+                }
```
It looks like we have both the data-object pattern, and array shaped data format. Can we pick one API? This is net new code, we shouldn't start off with two different APIs.

To me the data-object pattern is easier to operate as I can use LSP to explore it more easily. It is also less ambiguous and easier to evolve and extend.

dereuromark replied:
I am fine with only factories and objects.
We can then remove all array configs.

markstory commented on src/Datasource/Paging/SortField.php:
```diff
@@ -0,0 +1,140 @@
+/**
+ * Represents a sort field configuration for pagination.
+ */
+class SortField
```
I think classes like this are a great way for use to provide stronger typing and easier to operate and remember APIs.

markstory commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -433,7 +437,15 @@ public function sort(string $key, array|string|null $title = null, array $option
+        } else {
+            // Use traditional separate format
+            $paging = ['sort' => $key, 'direction' => $dir, 'page' => 1];
+        }
```
Do we need this if we're introducing `SortField`?

markstory commented on tests/TestCase/Datasource/Paging/SortMapFactoryTest.php:
```diff
@@ -0,0 +1,244 @@
+     */
+    public function testFluentInterface(): void
+    {
+        $sortMap = SortMapFactory::create()
+            ->sortKey('newest')
+                ->desc('created')
+                ->asc('title')
+            ->sortKey('oldest')
+                ->asc('created')
+                ->asc('title')
+            ->sortKey('popular')
+                ->locked('score', SortField::DESC)
+                ->desc('views')
+            ->sortKey('alphabetical')
+                ->asc('name')
+            ->build();
```
Do we need two different builder patterns? Can we pick one instead.

dereuromark replied:
Which ones u prefer?

josbeir replied:
1. i like the SortMapFactory,, i'm just not sure about the fluent style of building. This is something that is not used in other cake api's i think and seems a bit too locked in.

2. I'd additionally love to see something that would expose an instance of the factory through a callable, this way you would not need to call ->build() 

```php
'sortMap' => function(Builder $builder) {
  return $builder
      ->key('newest', SortField::desc('title'), SortField::asc('title'))
      ->key('oldest', SortField::asc('created'), SortField::asc('title'))
      ->key('popular',
            SortField::desc('score', locked: true),
            SortField::desc('view'),
           ....
           ....
     ); 
});
```

3. i would not use the word 'sort' (sortKey) in method names, This seems a bit redundant given the Factory shares the same name :-) 

4.  Maybe we should just use use `'sorts` as key to define the sortMap ?

markstory replied:
> Which ones u prefer?

I find the factory easier to understand when reading code. The fluent builder is interesting, but the method chains are dense, and I could see a stray sortKey() call causing unexpected results.

dereuromark replied:
So only keeping SortField?
I also like the idea of SortMapFactory and maybe a callable support.

rochamarcelo replied:
I would keep the SortField and SortMap (with a  set method). SortMap would not be a factory  class but hold the real map. 

Maybe move the resolveSortMapping logic to SortMap class, to allow custom logic

dereuromark replied:
Do you want to make a PR with the final showcase/api of this targeting this one?

josbeir replied:
I believe we should decide on one of the following options:

1. SortField only
2. Fluent Builder + `SortField` (as Enum/Consts - `SortField::DESC`)
3. Factory builder + typed `SortField` (`SortField::asc('fieldName')`
4. Same as 3 + callable shortcut; also opens up options for clean organization

In case of factory builder: do pass sortFields as an array or as variadic parameters.

Personally i'd go for  option 4 + variadic

rochamarcelo replied:
Here is my idea https://github.com/cakephp/cakephp/pull/18924

Sample code

```php
    $map = (new SortMap())
      ->push('active',
          new SortField("active"),
          "name",
          SortField::locked("created", SortField::DESC),
      )
      ->push('title')
      ->push("group", "Groups.name");
```
Let me know what you think.

dereuromark replied:
@josbeir Do you want to make a PR towards here with your suggested improvements?

josbeir replied:
> @josbeir Do you want to make a PR towards here with your suggested improvements?

I'll try to take a shot at it in the coming days

josbeir committed:
Refactor pagination sorting: rename sortMap to sorts with callable factory

- Rename 'sortMap' configuration option to 'sorts'
- Rename SortMapFactory to SortsFactory with non-fluent API
- Remove SortField::locked() method, add $locked parameter to asc()/desc()
- Add callable 'sorts' option that receives SortsFactory instance

The new API supports:
```php
'sorts' => function($factory) {
    return $factory
        ->add('name', SortField::asc('Users.name'))
        ->add('popularity', SortField::desc('score', locked: true));
}
```

josbeir committed:
Cs fix

josbeir committed:
More cs fixes

josbeir commented:
Updated with following changes:

- Rename sortMap → sorts configuration option
- Rename SortMapFactory → SortsFactory with simplified non-fluent API
- Replace SortField::locked() with parameter: SortField::desc('field', locked: true)
- Add callable sorts option with add() method accepting unlimited SortField arguments
- Maintain backward compatibility with deprecation warnings

The current thought i have: do we keep `SortFieldFactory`, `SortsFactory` or both. I think they both could be useful.

### Current implementation covers:

Basic:
```php
$settings = [
    'sorts' => [
        'User.id', // Use field as is.
        'name' => 'Users.name',
        'email' => 'Users.email',
        'newest' => [
            'created' => 'desc!',  // Always DESC - locked
            'title' => 'asc',      // Default ASC, but toggleable
        ],
        'created' => [
           'Users.created',
           'User.name'
         ],
   ]
];
```

SortField Objects:
```php
$settings = [
    'sorts' => [
        'newest' => [
            SortField::desc('created'),  // Default desc, toggleable
            SortField::asc('title'),     // Secondary sort
        ],
        'popular' => [
            SortField::desc('score', locked: true), // Always desc
            'author',  // Still support strings for BC
        ],
        'alphabetical' => SortField::asc('title'), // Single field
        'priority' => SortField::desc('priority', locked: true), // Locked
    ]
];
```

Factory builder
```php
// Simple
$settings = [
    'sorts' => [
        'relevance' => SortFieldFactory::create()
            ->desc('search_score', locked: true)  // Always desc
            ->desc('popularity')                  // Default desc, toggleable
            ->asc('title')                       // Tiebreaker
            ->build(),
            
        'price_low' => SortFieldFactory::create()
            ->asc('price', locked: true)         // Always ascending
            ->asc('title')                       // Tiebreaker
            ->build(),
            
        'price_high' => SortFieldFactory::create()
            ->desc('price', locked: true)        // Always descending
            ->asc('title')                       // Tiebreaker
            ->build(),
    ]
];

// Instance
$factory = new SortsFactory();
$factory->add('newest', SortField::desc('published'), SortField::asc('title'));
$factory->add('popular', SortField::desc('views', locked: true), SortField::desc('comments'));
$factory->add('alphabetical', SortField::asc('title'));
$factory->add('author');  // Shorthand: creates 'author' => ['author']

$settings = [
    'sorts' => $factory->build()
];

// Callable with new instance, should return instance w/o build() method
$settings = [
    'sorts' => function(SortsFactory $factory) {
        return $factory
            ->add('newest', 
                SortField::desc('created'), 
                SortField::asc('title')
            )
            ->add('popular', 
                SortField::desc('score', locked: true),
                SortField::desc('comments')
            )
            ->add('alphabetical', SortField::asc('title'));
    }
];
```

josbeir committed:
Merge remote-tracking branch 'origin/5.next' into 5.next-paginator-sort-map

ADmad commented on src/Datasource/Paging/NumericPaginator.php:
```diff
@@ -49,12 +49,26 @@ class NumericPaginator implements PaginatorInterface
      *   have to explicitly specify them (along with other fields). Using an empty
      *   array will disable sorting alltogether.
+     *
+     * @deprecated 5.3.0 Use `sorts` with SortField objects instead for more flexible sorting configuration.
```
Can an existing config like `'sortableFields' => ['title', 'status']` be simply replaced by `'sorts' => ['title', 'status']`?

josbeir replied:
@ADmad  i have updated the examples [here](https://github.com/cakephp/cakephp/pull/18898#issuecomment-3391934637) to cover current implemented functionality, probably a bit clearer to follow progress using that as reference

ADmad commented:
The existing tests cases using the deprecated `sortableFields` option will need to be wrapped in `$this->deprecated(..)` to fix CI.

Since the new `sorts` option also supports a list of fields as used by `sortableFields`, couldn't be add the new feature using the same option name `sortableFields`?  That way existing apps won't require any change and one can update `sortableFields` are required to take advance of the new "sort field" field feature.  Option name `sorts` sounds a bit weird to me.

ADmad commented on src/Datasource/Paging/SortsFactory.php:
```diff
@@ -0,0 +1,59 @@
+ *
+ * Provides non-fluent interface for building sorts with multiple sort keys and fields.
+ */
+class SortsFactory
```
Not too thrilled about this class name. Assuming we continue using the option name `sortableFields` as I suggested in another comment, this class could be named `SortableFieldsFactory`.

ADmad replied:
More nitpick, should we use the `Builder` suffix instead of `Factory`, to be consistent with other classes like  `RoutesBuidler`, `ViewBuilder`, `CorsBuilder`, `FunctionsBuilder`.

josbeir replied:
Added to [plan](https://github.com/cakephp/cakephp/pull/18898#issuecomment-3393089013)

josbeir commented:
> The existing tests cases using the deprecated `sortableFields` option will need to be wrapped in `$this->deprecated(..)` to fix CI.
> 
> Since the new `sorts` option also supports a list of fields as used by `sortableFields`, couldn't be add the new feature using the same option name `sortableFields`? That way existing apps won't require any change and one can update `sortableFields` are required to take advance of the new "sort field" field feature. Option name `sorts` sounds a bit weird to me.

So, current plan (based on comments below):

* Merge new functionality into existing `sortableFields` option
* Refactor `SortsFactory` to `SortableFieldsBuilder`
* Refactor `SortFieldFactory` to `SortFieldBuilder`
* Move [map resolving](https://github.com/cakephp/cakephp/pull/18898#issuecomment-3394504445) to `SortableFieldsBuilder` instead of `NumericPaginator::resolveSortMapping`

No need for $this->deprecated(..) in tests then :-) 

Things to keep in mind:
- https://github.com/cakephp/cakephp/pull/18898#discussion_r2350555774 (only use OOP for building)
- https://github.com/cakephp/cakephp/pull/18898#discussion_r2350569893 (do we need 2 different Factories)
- https://github.com/cakephp/cakephp/pull/18898#discussion_r2350566947 (do we need configurable `options.sortFormat`)

ADmad commented:
> So new plan:
>
>    Merge new functionality into existing sortableFields option
>    Refactor SortsFactory to SortableFieldsFactory
>
> No need for $this->deprecated(..) in tests then :-)

Works for me (with `SortableFieldsBuilder`), but please wait for feedback from others :)

rochamarcelo commented:
Should I close https://github.com/cakephp/cakephp/pull/18924 ? 

What do you think about the changes in there?

dereuromark commented:
Builder sounds also fine to me.

josbeir commented:
> Should I close #18924 ?
> 
> What do you think about the changes in there?

@rochamarcelo I think the new SortsFactory behaves exactly the same way as your  idea now, The difference is more in the way the SortField class is used. I *personally* like the explicitness of how it is implemented now:

```php
SortField::desc('score', locked: true),
SortField::asc('comments')
```

rochamarcelo commented:
The idea there is to move logic to SortMap, this allow application to customize how the map should work, and remove complexity from Paginator class. No changes to SortField

josbeir commented:
> The idea there is to move logic to SortMap, this allow application to customize how the map should work, and remove complexity from Paginator class. No changes to SortField

This is the case now right?  Please check examples here: https://github.com/cakephp/cakephp/pull/18898#issuecomment-3391934637

rochamarcelo commented:
> > The idea there is to move logic to SortMap, this allow application to customize how the map should work, and remove complexity from Paginator class. No changes to SortField
> 
> This is the case now right? Please check examples here: [#18898 (comment)](https://github.com/cakephp/cakephp/pull/18898#issuecomment-3391934637)

I'm talking about https://github.com/cakephp/cakephp/pull/18924/files#diff-5fa213cf6863d25393080833a687bb1aa3d8d83209a8206889a48c8acabc7221R60  moved code from NumericPaginator::resolveSortMapping to SortMap::resolve and made logic simpler.

josbeir commented:
> > > The idea there is to move logic to SortMap, this allow application to customize how the map should work, and remove complexity from Paginator class. No changes to SortField
> > 
> > 
> > This is the case now right? Please check examples here: [#18898 (comment)](https://github.com/cakephp/cakephp/pull/18898#issuecomment-3391934637)
> 
> I'm talking about https://github.com/cakephp/cakephp/pull/18924/files#diff-5fa213cf6863d25393080833a687bb1aa3d8d83209a8206889a48c8acabc7221R60 moved code from NumericPaginator::resolveSortMapping to SortMap::resolve and made logic simpler.

:heavy_check_mark: i understand now, added it to the plan!

josbeir committed:
Merge remote-tracking branch 'origin/5.next' into 5.next-paginator-sort-map

josbeir committed:
Refactoring Builder suffix + sortableFields option

- Factory -> Builder suffix
- Favor `sortableFields` instead of `sorts` or `sortMap`

josbeir committed:
Refactor SortableFieldsBuilder: Move

- Moved resolving of sort array to SortableFieldsBuilder::resolve
instead of  `NumericPaginator::resolveSortMapping()`
- Remove previously introduced '!' locking syntax in favor of
SortField::desc()/asc() with locked parameter
- Drop previously introduced combined sort filters

josbeir commented:
Implemented plan from https://github.com/cakephp/cakephp/pull/18898#issuecomment-3393089013

* Refactored "Factory" to "Builder" suffix
* Favor `sortableFields` instead of `sorts` or `sortMap` preserving BC
* Removed legacy '!' syntax that @dereuromark introduced early in this PR in favor of SortField `locked` argument.
* Moved resolving of sort array to `SortableFieldsBuilder::resolve` instead of  `NumericPaginator::resolveSortMapping()`

josbeir committed:
CS fix

josbeir committed:
drop: fluent SortFieldBuilder introduced early in this pr

ADmad commented on src/Datasource/Paging/NumericPaginator.php:
```diff
@@ -668,6 +706,28 @@ protected function _prefix(RepositoryInterface $object, array $order, bool $allo
+            $direction = 'asc';
+        }
+
+        return [$sortKey, $direction, $directionSpecified];
```
Can we return an associative array instead, that's more readable?

ADmad commented on src/Datasource/Paging/SortableFieldsBuilder.php:
```diff
@@ -0,0 +1,224 @@
+     * @param callable|array<mixed>|null $config The sortableFields configuration
+     * @return self|null Builder instance or null if no config
+     */
+    public static function create(array|callable|null $config): ?self
```
The `callable` can be changed to `Closure`, that would suffice for most use cases. Callable checks are expensive in PHP.

ADmad commented on src/Datasource/Paging/SortableFieldsBuilder.php:
```diff
@@ -0,0 +1,224 @@
+     * @param array<mixed> $config Array configuration
+     * @return self
+     */
+    public static function fromArray(array $config): self
```
If we are going to use `self` return type then maybe make the class `final`? Or else change the return to `static`.

ADmad commented on src/Datasource/Paging/SortableFieldsBuilder.php:
```diff
@@ -0,0 +1,224 @@
+     *
+     * @return array<string, array<\Cake\Datasource\Paging\SortField|string>>
+     */
+    public function build(): array
```
There's no building process inside this method, it's just a getter :)

ADmad commented on src/Datasource/Paging/NumericPaginator.php:
```diff
@@ -668,6 +706,28 @@ protected function _prefix(RepositoryInterface $object, array $order, bool $allo
+        $directionSpecified = isset($options['direction']);
+
+        // Validate direction
+        if (!in_array($direction, ['asc', 'desc'], true)) {
```
`SortField` now has constants for `asc`, `desc`.

ADmad commented on src/Datasource/Paging/SortField.php:
```diff
@@ -0,0 +1,130 @@
+    {
+        $this->field = $field;
+        $this->defaultDirection = $defaultDirection;
+        $this->locked = $locked;
```
We can use constructor property promotion instead of these assignments.

rochamarcelo commented on src/Datasource/Paging/NumericPaginator.php:
```diff
@@ -539,24 +564,49 @@ protected function getDefaults(string $alias, array $settings): array
     protected function validateSort(RepositoryInterface $object, array $options): array
     {
+        // Check if we have sortableFields configured
+        $builder = SortableFieldsBuilder::create($options['sortableFields'] ?? null);
```
sortableFields Could accept a SortableFieldsBuilder object

rochamarcelo commented on src/Datasource/Paging/SortableFieldsBuilder.php:
```diff
@@ -0,0 +1,224 @@
+                    $builder->map[$value] = [$value];
+                } else {
+                    // String key: use as-is
+                    $builder->map[$key] = $value;
```
Maybe call `$this->add` to make sure it has correct types?

josbeir committed:
Address review feedback from #3353455462

- Make SortableFieldsBuilder final with self:: calls
- Use Closure instead of callable for performance
- Rename build() to toArray()
- Use associative arrays in parseSortParams() for readability
- Replace direction string literals with SortField constants

josbeir committed:
Refactor SortableFieldsBuilder to use type-safe set() method and support builder instances in NumericPaginator

josbeir commented:
Thanks for these reviews @ADmad and @rochamarcelo . Updated with your proposed changes!

josbeir commented on src/Datasource/Paging/SortableFieldsBuilder.php:
```diff
@@ -130,10 +132,39 @@ public function add(string $sortKey, SortField|string ...$fields)
+            $this->map[$sortKey] = $value;
+        } else {
+            // Fallback for other types
+            $this->map[$sortKey] = $value;
```
Should we allow this or throw an exception ?

rochamarcelo replied:
I think so., but maybe can reduce the code with 

```
$value = is_array($value) ? $value : [$value];
$this->add($sortKey, ...$value);
```

rochamarcelo replied:
Or one line: `$this->add($sortKey, ...is_array($value) ? $value : [$value]);`

josbeir replied:
Added the exception, we still need to do type checking for legacy 'key' => 'value' support.
I could look into casting these things to an array but that would require adjusting a bunch of tests

josbeir committed:
Add more tests

ADmad commented on src/Datasource/Paging/NumericPaginator.php:
```diff
@@ -539,24 +564,58 @@ protected function getDefaults(string $alias, array $settings): array
+            $builder = $sortableFields;
+        } else {
+            $builder = SortableFieldsBuilder::create($sortableFields);
+        }
```
```suggestion
        $builder = $sortableFields instanceof SortableFieldsBuilder
            ? $sortableFields
            : SortableFieldsBuilder::create($sortableFields);
```

ADmad commented on src/Datasource/Paging/SortField.php:
```diff
@@ -0,0 +1,115 @@
+    public function getDirection(string $requestedDirection, bool $directionSpecified): string
+    {
+        if ($this->locked) {
+            return $this->defaultDirection ?? static::ASC;
```
Other usages of the constants in the class use `self::`.

josbeir committed:
Cleanup

ADmad reviewed (approved):
Needs a corresponding docs PR.

dereuromark commented:
What happened with our key reason to add this?

> Imo we dont even need even sort asc/desc if we bind that into the key.
> sort=title-asc
> sort=title-desc
> A flat list with all possible options

Those combined fields that become a flat list to toggle, and as such allow multiple fields to be in one single sort field key?
I guess thats out of scope for now.

So all good.

josbeir commented:
> What happened with our key reason to add this?
> 
> > Imo we dont even need even sort asc/desc if we bind that into the key.
> > sort=title-asc
> > sort=title-desc
> > A flat list with all possible options
> 
> Those combined fields that become a flat list to toggle, and as such allow multiple fields to be in one single sort field key? I guess thats out of scope for now.
> 
> So all good.

We could indeed introduce this into a separate PR. I think there was no real consensus about this. Was looking at your change https://github.com/cakephp/cakephp/pull/18898/commits/2a5410571343fd9583862778c8bc576627bd61e3 and didn't see a real advantage except for maybe cosmetics. And maybe a potential SEO disadvantage when using the combined format as crawlers could see this as a separate page instead.

dereuromark commented:
It has a huge Improvement actually
A) Single field reduced complexity overall from the outside
B) supports multi column sort with asc for one and desc for the other (current solution cannot right?)
C) Allows more easily to have multiple paginations per URL (sort1, sort2, sort3, without the complexity of having to deal with matching dir keys as well) in the future

But yeah
I can make a feature flagged separate PR.

I dont think crawlers have more issues with that the current 2 keys.

dereuromark commented:
One real life example:
```
  ORDER BY 
      in_stock DESC,     -- Show in-stock items first
      price ASC,         -- Cheapest first (within stock status)
      rating DESC        -- Highest rated (within same price)
```      
If you use the flat (single combined field) approach of x-asc, x-desc, you can hide those behind it, as it will toggle those based on the keys here, not the inside asc/desc values of each of those based on the 2nd param.

  Auto-generates:
  - ?sort=best-deal-asc → ORDER BY in_stock DESC, price ASC, rating DESC
  - ?sort=best-deal-desc → ORDER BY in_stock ASC, price DESC, rating ASC (all inverted)

Or am I missing something and that is already possible with the separate fields?

That is basically what I have been missing for those 11+ years ;)

dereuromark commented:
Made the PR here to showcase it: https://github.com/cakephp/cakephp/pull/18981
If everyone wants to keep the separated one, ok.
I personally think a single field could be simplifying things, and it is opt-in only for now.

josbeir commented:
> One real life example:
> 
> ```
>   ORDER BY 
>       in_stock DESC,     -- Show in-stock items first
>       price ASC,         -- Cheapest first (within stock status)
>       rating DESC        -- Highest rated (within same price)
> ```
> 
> If you use the flat (single combined field) approach of x-asc, x-desc, you can hide those behind it, as it will toggle those based on the keys here, not the inside asc/desc values of each of those based on the 2nd param.
> 
> Auto-generates:
> 
>     * ?sort=best-deal-asc → ORDER BY in_stock DESC, price ASC, rating DESC
> 
>     * ?sort=best-deal-desc → ORDER BY in_stock ASC, price DESC, rating ASC (all inverted)
> 
> 
> Or am I missing something and that is already possible with the separate fields?
> 
> That is basically what I have been missing for those 11+ years ;)

This makes sense to me now, it is actually the inverting behavior that is key here right? Not the format?

Consider this map setup:

```php
'best-deal' => [
    SortField::desc('published'),  // default: desc
    SortField::asc('title'),       // default: asc
    SortField::desc('author_id'),  // default: desc
],
```

Current behavior:

- `?sort=best-deal` (no direction)` → Uses defaults (desc, asc, desc)
- `?sort=best-deal&direction=asc` → **ALL become asc** (asc, asc, asc)
- `?sort=best-deal&direction=desc` → **ALL become desc** (desc, desc, desc)

Expected behavior

- `?sort=best-deal&direction=asc` → Uses defaults (desc, asc, desc)
- `?sort=best-deal&direction=desc` → Inverts all (asc, desc, asc)

I think the later one should be default.

@dereuromark for me you can merge your PR back in but the inverting behavior should (and could?) also work on the current query string format (with sort & direction keys) right?

rochamarcelo commented:
About the builder class, why not make the $map property always have list of SortField to make the resolve and create methods simpler? 
The create method should be able to 'parse' correctly.

rochamarcelo commented:
I'm not sure why the buider class is final, I wish to be able to extend this class in application or maybe have an interface to allow custom class.

dereuromark commented:
I also think final is not needed here either way.

josbeir committed:
SortablefieldsBuilder: drop final class, introduce exception on invalid types.
- Don't use final for SortablefieldsBuilder, use lsb for methods.
- Add exception for invalid type in SortablefieldsBuilder::set()

ADmad commented:
> I'm not sure why the buider class is final, I wish to be able to extend this class in application or maybe have an interface to allow custom class.

I had suggested making the class final, since many of its methods have return type `self` extending would have been problematic. I should have probably suggested to use `static` instead of `self`.

josbeir committed:
Add unsafe static rule for SortableFieldsBuilder to phpstan.baseline

josbeir commented:
Dropped the final class, i was a bit to eager on this one :-)

dereuromark committed:
Paginator: Implement combined sorting key. (#18981)

* Implement combined sorting.

dereuromark committed:
Merge branch '5.next' into 5.next-paginator-sort-map

josbeir committed:
Clarify docblock

othercorey commented:
@josbeir @ADmad Is this version approved?

ADmad commented:
> Is this version approved?

I think it is. Wait for a couple of days for josbeir to return from vacation.

dereuromark commented:
He did add the last commit. So I assume he wasnt seeing any troubling things anymore.

josbeir commented:
For me this is now ok. @rochamarcelo ? @dereuromark ?

othercorey commented:
Feel free to merge. Don't know the details for the migration docs.

dereuromark commented:
We can still revise as we go.
I can whip up the docs.
